test/integration/consul-container/test/wanfed/acl_bootstrap_replication_test.go

```diff
@@ -0,0 +1,157 @@
+// SPDX-License-Identifier: MPL-2.0
+
+package wanfed
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/sdk/testutil/retry"
+	libcluster "github.com/hashicorp/consul/test/integration/consul-container/libs/cluster"
+)
+
+const replicationPolicyRules = `
+acl      = "write"
+operator = "write"
+
+service_prefix "" {
+  policy     = "read"
+  intentions = "read"
+}
+`
+
+var retryFuncTimer = &retry.Timer{
+	Timeout: 60 * time.Second,
+	Wait:    2 * time.Second,
+}
+
+// retryFunc is a helper to retry the given function until it returns a nil
+// error. It returns the value from the function on success.
+func retryFunc[T any](t *testing.T, f func() (T, error)) T {
+	var result T
+	retry.RunWith(retryFuncTimer, t, func(r *retry.R) {
+		val, err := f()
+		require.NoError(r, err)
+		result = val
+	})
+	return result
+}
+
+// TestWanFed_ReplicationBootstrap checks the setup procedure of two federated
+// datacenters with a manual ACL bootstrap step and ACL token replication enabled.
+//
+//   - WAN join two clusters that are not ACL bootstrapped
+//     and have ACL replication enabled
+//   - ACL Bootstrap the primary datacenter
+//   - Configure the replication token in each datacenter
+//   - Validate ACL replication by creating a token
+//     and checking it is found in each datacenter.
+//
+// Added to validate this issue: https://github.com/hashicorp/consul/issues/16620
+func TestWanFed_ReplicationBootstrap(t *testing.T) {
+	cfgFunc := func(c *libcluster.ConfigBuilder) {
+		c.Set("primary_datacenter", "primary")
+		c.Set("acl.enabled", true)
+		c.Set("acl.default_policy", "deny")
+		c.Set("acl.enable_token_persistence", true)
+		c.Set("acl.enable_token_replication", true)
+	}
+
+	_, primary := createNonACLBootstrappedCluster(t, "primary", cfgFunc)
+	_, secondary := createNonACLBootstrappedCluster(t, "secondary", func(c *libcluster.ConfigBuilder) {
+		cfgFunc(c)
+		c.Set("retry_join_wan", []string{primary.GetIP()})
+	})
+
+	// ACL bootstrap the primary
+	rootToken := retryFunc(t, func() (*api.ACLToken, error) {
+		t.Logf("ACL bootstrap the primary datacenter")
+		tok, _, err := primary.GetClient().ACL().Bootstrap()
+		return tok, err
+	})
+
+	agents := []libcluster.Agent{primary, secondary}
+
+	for _, agent := range agents {
+		// Make a new client with the bootstrap token.
+		_, err := agent.NewClient(rootToken.SecretID, true)
+		require.NoError(t, err)
+
+		t.Logf("Wait for members in %s datacenter", agent.GetDatacenter())
+		libcluster.WaitForMembers(t, agent.GetClient(), 1)
+	}
+
+	// Create and set the replication token
+	replicationPolicy := retryFunc(t, func() (*api.ACLPolicy, error) {
+		t.Logf("Create the replication policy")
+		p, _, err := primary.GetClient().ACL().PolicyCreate(
+			&api.ACLPolicy{
+				Name:  "consul-server-replication",
+				Rules: replicationPolicyRules,
+			},
+			nil,
+		)
+		return p, err
+	})
```

Comment:
I saw there was a helper to [create an agent token](https://github.com/hashicorp/consul/blob/cd80ea18ff2393b31eb689dbf6118a110d3390de/test/integration/consul-container/libs/cluster/cluster.go#L259-L269). I could move this to a helper alongside that?